server/src/main/java/io/crate/execution/ddl/tables/AddColumnTask.java

```diff
@@ -123,34 +112,52 @@ public ClusterState execute(ClusterState currentState, AddColumnRequest request)
+
+    /**
+     * Adds missing parent columns.
+     * Also tries to update request by taking existing references from the cluster state.
+     * @return NULL if all columns already exist.
+     */
+    @Nullable
+    static AddColumnRequest normalizeColumns(AddColumnRequest request, DocTableInfo currentTable) {
         List<Reference> newColumns = new ArrayList<>();
-        boolean anyMissing = false;
-        for (var newColumn : request.references()) {
-            if (newColumn.column().isTopLevel()) {
-                newColumns.add(newColumn);
+
+        // Get existing references
+        boolean allExist = true;
+        for (Reference ref : request.references()) {
+            Reference exists = currentTable.getReference(ref.column());
+            if (exists == null) {
+                allExist = false;
+                newColumns.add(ref); // Actually new column, take it from request.
             } else {
-                ColumnIdent parent = newColumn.column();
-                while ((parent = parent.getParent()) != null) {
-                    if (!Symbols.containsColumn(request.references(), parent)
-                            && !Symbols.containsColumn(newColumns, parent)) {
-                        newColumns.add(currentTable.getReference(parent));
-                        anyMissing = true;
-                    }
+                if (ref.valueType().id() != exists.valueType().id()) {
+                    throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                        "Column `%s` already exists with type `%s`. Cannot add same column with type `%s`",
+                        ref.column(),
+                        exists.valueType().getName(),
+                        ref.valueType().getName()));
                 }
-                newColumns.add(newColumn);
+                newColumns.add(exists); // Get updated reference from the cluster state.
             }
         }
-        if (anyMissing) {
-            return new AddColumnRequest(
-                request.relationName(),
-                newColumns,
-                request.checkConstraints(),
-                request.pKeyIndices()
-            );
-        } else {
-            return request;
+        if (allExist) {
+            return null;
         }
+
+        // Add missing parents
+        for (var newColumn : request.references()) {
+            ColumnIdent parent = newColumn.column();
+            while ((parent = parent.getParent()) != null) {
+                if (!Symbols.containsColumn(request.references(), parent)
+                        && !Symbols.containsColumn(newColumns, parent)) {
+                    newColumns.add(currentTable.getReference(parent));
+                }
+            }
+        }
+
+        // If allExist was not short-circuited, we have at least one updated reference, have to update request regardless of missing parents presence.
+        request.references(newColumns);
```

Comment:
Could we instead change the method signature to return the new columns instead of mutating them on the request? Mutations make it harder to reason about the code as its more difficult to keep track of where changes happen. 

We generally avoid it,  unless it's for performance and then we try to keep it encapsulated.

Reply:
Done in https://github.com/crate/crate/pull/14142/commits/f5095477d1c5e0faff479456f5e69f57c0b4fbfe